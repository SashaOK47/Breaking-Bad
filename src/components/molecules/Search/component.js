import PropTypes from "prop-types";
import style from "./styles.module.scss";
import { Link } from "react-router-dom";

const Search = ({
  handleSubmit,
  handleChangeInput,
  inputValue,
  searchCharacters,
  isOpen,
}) => {
  return (
    <div className={style.search}>
      <form onSubmit={handleSubmit}>
        <input
          type="text"
          placeholder="Поиск"
          onChange={handleChangeInput}
          value={inputValue}
        />

        <button type="submit">
          <svg
            width="17"
            height="17"
            viewBox="0 0 17 17"
            fill="none"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              fillRule="evenodd"
              clipRule="evenodd"
              d="M13 7.5C13 10.5376 10.5376 13 7.5 13C4.46243 13 2 10.5376 2 7.5C2 4.46243 4.46243 2 7.5 2C10.5376 2 13 4.46243 13 7.5ZM12.0491 13.4633C10.7873 14.4274 9.21054 15 7.5 15C3.35786 15 0 11.6421 0 7.5C0 3.35786 3.35786 0 7.5 0C11.6421 0 15 3.35786 15 7.5C15 9.21054 14.4274 10.7873 13.4633 12.0491L16.7071 15.2929L15.2929 16.7071L12.0491 13.4633Z"
              fill="#F1B926"
            />
          </svg>
          <span>Найти</span>
        </button>
        <ul className={style.list}>
          {inputValue && isOpen
            ? searchCharacters.map((character) => (
                <li className={style.item} key={character.char_id}>
                  <Link
                    className={style.link}
                    to={`/character/${character.char_id}`}
                  >
                    <img
                      className={style.image}
                      src={character.img}
                      alt={character.name}
                    />
                    <span className={style.name}>{character.name}</span>
                  </Link>
                </li>
              ))
            : null}
        </ul>
      </form>
    </div>
  );
};
Search.propTypes = {
  handleSubmit: PropTypes.func,
  handleChangeInput: PropTypes.func,
  inputValue: PropTypes.string,
  searchCharacters: PropTypes.array,
  isOpen: PropTypes.bool,
};
export default Search;
