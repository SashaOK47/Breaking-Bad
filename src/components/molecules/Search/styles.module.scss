@import "../../../assets/styles/main.scss";
.search {
  ::-webkit-scrollbar {
    border-radius: 9em;
  }
  ::-webkit-scrollbar-thumb {
    background-color: $colorBgLink;
  }
  margin-bottom: 5rem;
  @include mobile {
    margin-bottom: 3rem;
  }
  form {
    margin: 0 auto;
    max-width: 49rem;
    width: 100%;
    position: relative;
    input {
      width: 100%;
      padding: 1rem 7rem 1rem 1.5rem;
      background: $white;
      border-radius: 0.25rem;
      @include boxShadowCard;
      &::placeholder {
        @include fontSmall;
        opacity: 0.4;
      }
      &:focus-visible {
        @include boxShadowFocus;
      }
    }
    button {
      display: flex;
      align-items: center;
      background: none;
      position: absolute;
      right: 0;
      top: 50%;
      padding: 0.8rem 1.5rem;
      border-radius: 0.25rem;
      transform: translateY(-50%);
      transition: background 0.5s;
      &:hover {
        background: $colorBgLink;
        svg path {
          fill: $mainFontColor;
        }
      }
      svg path {
        transition: fill 0.5s;
      }
      span {
        margin-left: 0.5rem;
        @include mainFont;
      }
      &:focus-visible {
        @include boxShadowFocus;
      }
    }
    .list {
      margin-top: 0.5rem;
      background: $white;
      @include boxShadowCard;
      border-radius: 0.25rem;
      position: absolute;
      width: 100%;
      max-height: 11.5rem;
      height: auto;
      overflow: auto;
      z-index: 1;
      .item {
        padding: 0.5rem;
        transition: background 0.3s;
        cursor: pointer;
        &:hover {
          background-color: darken($bgColor, 1%);
        }
        .link {
          display: flex;
          align-items: center;
          width: 100%;
          &:focus-visible {
            @include boxShadowFocus;
          }
          .image {
            width: 1.9rem;
          }
          .name {
            margin-left: 0.5rem;
          }
        }
      }
    }
  }
}
