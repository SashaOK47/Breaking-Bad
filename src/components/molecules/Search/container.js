import Search from "./component";
import { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import {
  setCharactersSearch,
  setSearchQuery,
} from "../../../store/actions/characters";

const SearchContainer = () => {
  const dispatch = useDispatch();
  const characters = useSelector((state) => state.characters.characters);

  const [inputValue, setInputValue] = useState("");
  const [isOpen, setIsOpen] = useState(true);

  let searchCharacters = [];

  searchCharacters = characters.filter((character) => {
    return character.name.toLowerCase().includes(inputValue.toLowerCase());
  });

  const handleChangeInput = (e) => {
    setIsOpen(true);
    setInputValue(e.target.value);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    if (!inputValue.trim()) return;
    dispatch(setCharactersSearch(searchCharacters));
    dispatch(setSearchQuery(inputValue));
    setInputValue("");
  };

  return (
    <Search
      handleSubmit={handleSubmit}
      handleChangeInput={handleChangeInput}
      inputValue={inputValue}
      searchCharacters={searchCharacters}
      isOpen={isOpen}
    />
  );
};
export const container = SearchContainer;
