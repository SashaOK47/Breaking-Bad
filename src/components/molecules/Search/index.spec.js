import React from "react";
import Search from "./component";

const handleChangeInput = jest.fn();
let inputValue = "";
const searchCharacters = [
  {
    char_id: 1,
    name: "Test",
    status: "test",
    img: "https://test.test",
    birthday: "12.12.12",
  },
  {
    char_id: 1,
    name: "Test2",
    status: "test2",
    img: "https://test2.test",
    birthday: "12.12.21",
  },
];
let isOpen = false;

const setUp = (props) => shallow(<Search {...props} />);

describe("Search component", () => {
  it("should render Search component", () => {
    const component = setUp({ searchCharacters, inputValue, isOpen });
    expect(component).toMatchSnapshot();
  });

  it("should render list block", () => {
    isOpen = true;
    inputValue = "Test";
    const component = setUp({
      searchCharacters,
      isOpen,
      inputValue,
    });
    expect(component).toMatchSnapshot();
  });

  it("worked out the method handleChangeInput", () => {
    const component = setUp({
      handleChangeInput,
    });
    component.find("input").simulate("change", { target: { value: "Test" } });
    expect(handleChangeInput).toHaveBeenCalled();
  });
});
