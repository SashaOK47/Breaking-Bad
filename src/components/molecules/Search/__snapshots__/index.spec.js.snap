// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component should render Search component 1`] = `
<div
  className="search"
>
  <form>
    <input
      placeholder="Поиск"
      type="text"
      value=""
    />
    <button
      type="submit"
    >
      <svg
        fill="none"
        height="17"
        viewBox="0 0 17 17"
        width="17"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clipRule="evenodd"
          d="M13 7.5C13 10.5376 10.5376 13 7.5 13C4.46243 13 2 10.5376 2 7.5C2 4.46243 4.46243 2 7.5 2C10.5376 2 13 4.46243 13 7.5ZM12.0491 13.4633C10.7873 14.4274 9.21054 15 7.5 15C3.35786 15 0 11.6421 0 7.5C0 3.35786 3.35786 0 7.5 0C11.6421 0 15 3.35786 15 7.5C15 9.21054 14.4274 10.7873 13.4633 12.0491L16.7071 15.2929L15.2929 16.7071L12.0491 13.4633Z"
          fill="#F1B926"
          fillRule="evenodd"
        />
      </svg>
      <span>
        Найти
      </span>
    </button>
    <ul
      className="list"
    />
  </form>
</div>
`;

exports[`Search component should render list block 1`] = `
<div
  className="search"
>
  <form>
    <input
      placeholder="Поиск"
      type="text"
      value="Test"
    />
    <button
      type="submit"
    >
      <svg
        fill="none"
        height="17"
        viewBox="0 0 17 17"
        width="17"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clipRule="evenodd"
          d="M13 7.5C13 10.5376 10.5376 13 7.5 13C4.46243 13 2 10.5376 2 7.5C2 4.46243 4.46243 2 7.5 2C10.5376 2 13 4.46243 13 7.5ZM12.0491 13.4633C10.7873 14.4274 9.21054 15 7.5 15C3.35786 15 0 11.6421 0 7.5C0 3.35786 3.35786 0 7.5 0C11.6421 0 15 3.35786 15 7.5C15 9.21054 14.4274 10.7873 13.4633 12.0491L16.7071 15.2929L15.2929 16.7071L12.0491 13.4633Z"
          fill="#F1B926"
          fillRule="evenodd"
        />
      </svg>
      <span>
        Найти
      </span>
    </button>
    <ul
      className="list"
    >
      <li
        className="item"
        key="1"
      >
        <Link
          className="link"
          to="/character/1"
        >
          <img
            alt="Test"
            className="image"
            src="https://test.test"
          />
          <span
            className="name"
          >
            Test
          </span>
        </Link>
      </li>
      <li
        className="item"
        key="1"
      >
        <Link
          className="link"
          to="/character/1"
        >
          <img
            alt="Test2"
            className="image"
            src="https://test2.test"
          />
          <span
            className="name"
          >
            Test2
          </span>
        </Link>
      </li>
    </ul>
  </form>
</div>
`;
